### PR TITLE
Add event in webhook

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve? 
Add event in webhook before the pod is admitted to delete. It would help us know the tidbcluster running process when pod admissionwebhook is enabled.

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

### Does this PR introduce a user-facing change?:

```release-note
Add useful `Event` in `TidbCluster` during upgrading and scaling when `admissionWebhook.validation.pods` in operator configuration is enabled.
```

### CODE REVIEW

File: pkg/webhook/admission_hooks.go [hunk 1]
```diff
@@ -148,7 +152,15 @@ func (a *AdmissionHook) Initialize(cfg *rest.Config, stopCh <-chan struct{}) err
...
 	// init pdControl
 	pdControl := pdapi.NewDefaultPDControl(kubeCli)
-	pc := pod.NewPodAdmissionControl(kubeCli, cli, pdControl, strings.Split(a.ExtraServiceAccounts, ","), a.EvictRegionLeaderTimeout)
+
+	//init recorder
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(klog.Infof)
+	eventBroadcaster.StartRecordingToSink(&eventv1.EventSinkImpl{
+		Interface: eventv1.New(kubeCli.CoreV1().RESTClient()).Events("")})
+	recorder := eventBroadcaster.NewRecorder(v1alpha1.Scheme, corev1.EventSource{Component: "tidbcluster"})
```
Comment (weekface) [inline pkg/webhook/admission_hooks.go:161]:
```suggestion
	recorder := eventBroadcaster.NewRecorder(v1alpha1.Scheme, corev1.EventSource{Component: "tidb-admission-controller"})
```
?

File: pkg/webhook/pod/pods.go [hunk 2]
```diff
@@ -44,17 +45,24 @@ type PodAdmissionControl struct {
...
 	// the map of the service account from the request which should be checked by webhook
 	serviceAccounts sets.String
+	// recorder to send event
+	recorder record.EventRecorder
 }
 
 const (
 	stsControllerServiceAccounts = "system:serviceaccount:kube-system:statefulset-controller"
+	podDeleteMsgPattern          = "pod[%s] deleted"
+	pdScaleInReason              = "pd scale-in"
```
Comment (weekface) [inline pkg/webhook/pod/pods.go:55]:
```suggestion
	pdScaleInReason              = "PDScaleIn"
```
?

Comment (Yisaer) [reply pkg/webhook/pod/pods.go]: updated.

File: pkg/webhook/pod/pods.go [hunk 3]
```diff
@@ -44,17 +45,24 @@ type PodAdmissionControl struct {
...
+	recorder record.EventRecorder
 }
 
 const (
 	stsControllerServiceAccounts = "system:serviceaccount:kube-system:statefulset-controller"
+	podDeleteMsgPattern          = "pod[%s] deleted"
+	pdScaleInReason              = "PDScaleIn"
+	pdUpgradeReason              = "PDUpgrade"
+	tikvScaleInReason            = "TiKVScaleIn"
+	tikvUpgradeReason            = "PDUpgrade"
```
Comment (DanielZhangQD) [inline pkg/webhook/pod/pods.go:58]:
```suggestion
	tikvUpgradeReason            = "TiKVUpgrade"
```

File: pkg/webhook/pod/pods.go [hunk 4]
```diff
@@ -44,17 +45,24 @@ type PodAdmissionControl struct {
 	pdControl pdapi.PDControlInterface
 	// the map of the service account from the request which should be checked by webhook
 	serviceAccounts sets.String
+	// recorder to send event
+	recorder record.EventRecorder
 }
 
 const (
 	stsControllerServiceAccounts = "system:serviceaccount:kube-system:statefulset-controller"
+	podDeleteMsgPattern          = "pod[%s] deleted"
```
Comment (DanielZhangQD) [inline pkg/webhook/pod/pods.go:54]: Add namespace?

File: pkg/webhook/pod/tikv_deleter.go [hunk 5]
```diff
@@ -144,6 +145,14 @@ func (pc *PodAdmissionControl) admitDeleteUselessTiKVPod(payload *admitPayload)
 		}
 	}
 
+	isUpgrading := operatorUtils.IsStatefulSetUpgrading(payload.ownerStatefulSet)
+	if isUpgrading {
+		// If the tikv is both during upgrading and scaling-in, we would take it as upgrading
```
Comment (DanielZhangQD) [inline pkg/webhook/pod/tikv_deleter.go:150]: Is it OK to generate two events for this case?

File: pkg/webhook/pod/pods.go (hunk 4 above)
Comment (DanielZhangQD) [inline pkg/webhook/pod/pods.go:54]:
```suggestion
	podDeleteMsgPattern          = "pod [%s] deleted"
```

File: pkg/webhook/pod/tikv_deleter.go (hunk 5 above)
Comment (Yisaer) [reply pkg/webhook/pod/tikv_deleter.go:150]: I suggest not as it is misleading.

File: pkg/webhook/pod/pods.go (hunk 4 above)
Comment (Yisaer) [reply pkg/webhook/pod/pods.go:54]: I think the namespace here is not necessary as the event belongs to a certain tidbcluster.

File: pkg/webhook/pod/tikv_deleter.go (hunk 5 above)
Comment (DanielZhangQD) [reply pkg/webhook/pod/tikv_deleter.go:150]: Normal upgrading should not be in this function, right?
There are other functions that upgrading may be hanled in, such as `admitDeleteUpTiKVPod`, etc.

File: pkg/webhook/pod/pods.go (hunk 4 above)
Comment (DanielZhangQD) [reply pkg/webhook/pod/pods.go]: Then add tc info?
Is the tc info already shown in the event?

File: pkg/webhook/pod/pd_deleter.go [hunk 6]
```diff
@@ -136,10 +141,10 @@ func (pc *PodAdmissionControl) admitDeleteNonPDMemberPod(payload *admitPayload)
 				klog.Infof("tc[%s/%s]'s pod[%s/%s] failed to update pvc,%v", namespace, tcName, namespace, name, err)
 				return util.ARFail(err)
 			}
+			pc.recorder.Event(payload.tc, corev1.EventTypeNormal, pdScaleInReason, podDeleteEventMessage(name))
```
Comment (DanielZhangQD) [inline pkg/webhook/pod/pd_deleter.go:144]: Need to add an event in L135?

File: pkg/webhook/pod/tikv_deleter.go [hunk 7]
```diff
@@ -144,6 +145,10 @@ func (pc *PodAdmissionControl) admitDeleteUselessTiKVPod(payload *admitPayload)
 		}
 	}
 
+	if !isInOrdinal {
+		pc.recorder.Event(payload.tc, corev1.EventTypeNormal, tikvScaleInReason, podDeleteEventMessage(name))
+	}
+
```
Comment (DanielZhangQD) [inline pkg/webhook/pod/tikv_deleter.go:151]: Move this block ahead of L146.

Comment (DanielZhangQD) [reply pkg/webhook/pod/tikv_deleter.go:151]: Event for L137?

File: pkg/webhook/pod/pd_deleter.go (hunk 6 above)
Comment (Yisaer) [reply pkg/webhook/pod/pd_deleter.go:145]: updated.

File: pkg/webhook/pod/tikv_deleter.go (hunk 5 above)
Comment (Yisaer) [reply pkg/webhook/pod/tikv_deleter.go]: updated.

File: pkg/webhook/pod/tikv_deleter.go (hunk 7 above)
Comment (Yisaer) [reply pkg/webhook/pod/tikv_deleter.go]: updated.

File: pkg/webhook/pod/tikv_deleter.go [hunk 8]
```diff
@@ -215,6 +218,7 @@ func (pc *PodAdmissionControl) admitDeleteUpTiKVPodDuringUpgrading(payload *admi
 		}
 	}
 
+	pc.recorder.Event(payload.tc, corev1.EventTypeNormal, tikvUpgradeReason, podDeleteEventMessage(name))
```
Comment (DanielZhangQD) [inline pkg/webhook/pod/tikv_deleter.go:221]: May also need to add an event in `admitDeleteDownTikvPod` for upgrading.

**Pull request merged**